Calendar data

antoine-de opened the pull request:
first branch for grid calendar.

bring calendar data managment and new api to explore the data.

It lacks the association between a route|stop schedule and a calendar.

azime committed:
ed: add calendar

azime committed:
kraken : add calendar

azime committed:
kraken : tests

azime committed:
ed : ajout des tables pour la fiche horaire par période

azime committed:
kraken : Déplacement du type enum "exception_type"

dans l'objet "exceptiondate"

azime committed:
ed : import des données fiche horaire par periode

azime committed:
ed : alimentation de la base de données

azime committed:
ed : import

azime committed:
ed : manage exception

antoine-de committed:
Jormungandr: add calendars api

first modification for the calendars api

azime committed:
ed: binarisation of grid calendar

azime committed:
ed : external_code for canalendar

antoine-de committed:
Jormungandr: first calendar api

first solution was dumb, it now seems to work better

antoine-de committed:
Kraken: protobuff conversion

antoine-de committed:
Kraken: ptref graph changes

antoine-de committed:
Jormungandr: protobuf grammar to displya calendars

antoine-de committed:
Jormungandr: protobuf grammar to displya calendars

azime committed:
jormungandr : Add link to calendar

azime committed:
kraken : relation between Line and Calendar

azime committed:
kraken : add function get in calendar

antoine-de committed:
Jormungandr: add calendar integration tests

hop first real integration test \o/

azime committed:
kraken : add message

antoine-de committed:
Jormungandr: dynamic load region name for test

antoine-de committed:
Jormungandr: calendar test improvement

antoine-de committed:
Jormungandr: add place holder in test urls

add place holders to be able to test url like
/v1/coverage/{regions.id}/calendars/{calendars.id}/lines/{lines.id}/
calendars

antoine-de committed:
Jormungandr: split unit tests and integration tests

ini inclusion problem if all tests are run together, so split of the
unit tests and integration tests in the cmake

antoine-de committed:
Jormungandr: small change after first review

kinnou02 commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -640,6 +640,174 @@ void TripPropertiesFusioHandler::handle_line(Data&, const csv_row& row, bool is_
     }
 }
+
+void HPeriodFusioHandler::init(Data&){
```
pourquoi HPeriod?

kinnou02 commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -640,6 +640,174 @@ void TripPropertiesFusioHandler::handle_line(Data&, const csv_row& row, bool is_
+        boost::gregorian::date begin_date(boost::gregorian::not_a_date_time);
+        boost::gregorian::date end_date(boost::gregorian::not_a_date_time);
+
+        try{
```
ça vaudrait le coup de factoriser un peu j'pense

kinnou02 commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -640,6 +640,174 @@ void TripPropertiesFusioHandler::handle_line(Data&, const csv_row& row, bool is_
+
+        try{
+            begin_date = boost::gregorian::from_undelimited_string(row[begin_c]);
+        }catch(boost::bad_lexical_cast ) {
```
idéalement catch les exception en const &

kinnou02 commented on source/ed/types.h:
```diff
@@ -58,6 +58,27 @@ struct JourneyPatternPointConnection: public Header {
     bool operator<(const JourneyPatternPointConnection &other) const;
 };
 
+struct Period : public Header {
```
quel est l'avantage par rapport à http://www.boost.org/doc/libs/1_39_0/doc/html/date_time/gregorian.html#date_time.gregorian.date_period

kinnou02 commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -640,6 +640,174 @@ void TripPropertiesFusioHandler::handle_line(Data&, const csv_row& row, bool is_
+    calendar->uri = row[id_c];
+    calendar->external_code = row[id_c];
+    calendar->name =  row[name_c];
+    calendar->week_pattern[0] = is_active(monday_c, row);
```
une ptite énum pour les jours ca serait plus clair par la suite

kinnou02 commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -640,6 +640,174 @@ void TripPropertiesFusioHandler::handle_line(Data&, const csv_row& row, bool is_
+            if(!date.is_not_a_date()){
+                navitia::type::ExceptionDate exception_date;
+                exception_date.date = date;
+                exception_date.type = static_cast<navitia::type::ExceptionDate::ExceptionType>(boost::lexical_cast<int>(row[type_c]));
```
les cast peuvent echouer

kinnou02 commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -640,6 +640,174 @@ void TripPropertiesFusioHandler::handle_line(Data&, const csv_row& row, bool is_
+        auto find_predicate = [&](const types::Line* line ) {
+            return (line->external_code == row[line_c]);
+        };
+        auto it = std::find_if(data.lines.begin(),
```
ça serait pas plus efficace d'utiliser GtfsData::line_map?

kinnou02 commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -640,6 +640,174 @@ void TripPropertiesFusioHandler::handle_line(Data&, const csv_row& row, bool is_
+                        "  file has more than one calendar and no id column");
+        throw InvalidHeaders(csv.filename);
+    }
+    ed::types::Calendar* calendar = new ed::types::Calendar();
```
j'ai comme l'impression que les calendar ne sont jamais delete, mais j'ai pas fini de tout lire

antoine-de replied:
sisi ils sont embarqués dans les ITER_NAVITIA_TYPE (la super macro ...)

kinnou02 commented on source/ed/ed_reader.cpp:
```diff
@@ -1072,6 +1078,73 @@ void EdReader::fill_origin_destinations(navitia::type::Data& data, pqxx::work& w
+        if (cal != nullptr){
+            navitia::type::ExceptionDate exception_date;
+            exception_date.date = bg::from_string(const_it["datetime"].as<std::string>());
+            exception_date.type = static_cast<navitia::type::ExceptionDate::ExceptionType>(const_it["exception_type_id"].as<int>());
```
idem les cast peuvent echouer

kinnou02 commented on source/ed/types.h:
```diff
@@ -58,6 +58,27 @@ struct JourneyPatternPointConnection: public Header {
+    std::vector<navitia::type::ExceptionDate> exceptions;
+
+    navitia::type::Calendar* get_navitia_type() const;
+    navitia::type::idx_t to_ulog() const{ return week_pattern.to_ulong();}
```
c'est ulong

kinnou02 commented on source/jormungandr/jormungandr/interfaces/v1/ResourceUri.py:
```diff
@@ -100,7 +100,19 @@ def wrapper(*args, **kwargs):
+                })
+            if collection in ['calendars']:
+                data['links'].append({
+                    "href": url_for("v1.lines.collection", **kwargs),
```
j'comprends pas bien ce que ça fait, donc il manque au minimum un commentaire

kinnou02 commented on source/jormungandr/jormungandr/interfaces/v1/ResourceUri.py:
```diff
@@ -100,7 +100,19 @@ def wrapper(*args, **kwargs):
+                    "rel": "calendars",
+                    "templated": templated
+                })
+            if collection in ['calendars']:
```
ca devrait être un "=="

kinnou02 commented on source/jormungandr/jormungandr/interfaces/v1/Uri.py:
```diff
@@ -382,6 +382,28 @@ def __init__(self):
+        """ Retrieves calendars"""
+
+        def __init__(self):
+            Uri.__init__(self, is_collection, "calendars")
```
faudrait plutot passer par super

kinnou02 commented on source/jormungandr/tests/app-saving.py:
```diff
@@ -1,25 +1,68 @@
+                    except AssertionError as error:
+                        logging.warn("url in error : assertion failed : " + error.message)
+
+        logging.info("%d tests serialized " % nb_test)
```
il vaut mieux privilégier format

kinnou02 commented on source/jormungandr/tests/app-saving.py:
```diff
@@ -1,25 +1,68 @@
+
+    def serialize(self):
+        logging.info("cleaning files")
+        #clean_dir(saving_directory)
```
c'est voulu le commentaire là?

kinnou02 commented on source/jormungandr/tests/app-saving.py:
```diff
@@ -1,25 +1,68 @@
+                        response = self.tester.get(url)
+                        # url can be in error (we might want to check invalid url)
+                        # but we log if it's the case
+                        if response.status_code != 200:
```
c'est un bon début, t'arrete pas en si bon chemin :)

antoine-de replied:
oui faudrait peut etre lancer nose en fin de serialisation meuh bon...

kinnou02 commented on source/jormungandr/tests/utils.py:
```diff
@@ -1,11 +1,66 @@
+place_holder_regexp = re.compile("\{(.*)\.(.*)\}")
+
+
+def is_place_holder(part):
```
une docstring serait sympa (avec doctest peut etre)

kinnou02 commented on source/sql/ed/01-tables.sql:
```diff
@@ -388,7 +388,54 @@ CREATE TABLE IF NOT EXISTS navitia.stop_time (
 
+CREATE TABLE IF NOT EXISTS navitia.week_pattern (
+    id BIGSERIAL PRIMARY KEY,
+	monday BOOLEAN NOT NULL,
```
pour info, ya ça aussi: http://www.postgresql.org/docs/9.1/static/datatype-bit.html

antoine-de replied:
je sais pas trop si c'est plus clair. tu préfères toi ?

kinnou02 commented on source/sql/ed/01-tables.sql:
```diff
@@ -388,7 +388,54 @@ CREATE TABLE IF NOT EXISTS navitia.stop_time (
+	sunday BOOLEAN NOT NULL
+);
+
+CREATE TABLE IF NOT EXISTS navitia.exception_type (
```
un jour faudra qu'on se décide entre table de référence et enum. j'ai pas d'avis tranché pour le moment

kinnou02 commented on source/type/pb_converter.cpp:
```diff
@@ -930,6 +932,46 @@ void fill_pb_object(const nt::VehicleJourney* vj, const nt::Data& data,
+        pbnavitia::ExceptionType type;
+        switch (excep.type) {
+        case nt::ExceptionDate::ExceptionType::add:
+            type = pbnavitia::ADD;
```
faut qu'on se décide entre majuscule et minuscule, mais la c'est complétement aléatoire

kinnou02 commented on source/type/type.h:
```diff
@@ -399,7 +401,31 @@ struct Connection: public Header, hasProperties{
 
+struct ExceptionDate {
+    enum class ExceptionType {
+        sub = 0,      // Suppression
```
anglais

kinnou02 commented on source/type/type.h:
```diff
@@ -399,7 +401,31 @@ struct Connection: public Header, hasProperties{
 typedef Connection<StopPoint>  StopPointConnection;
 
+struct ExceptionDate {
+    enum class ExceptionType {
```
une class enum dans une class ca risque d’être pénible:

```
navitia::type::ExceptionDate::ExceptionType::sub
```

kinnou02 commented on source/type/type.h:
```diff
@@ -566,6 +593,12 @@ struct JourneyPattern : public Header, Nameable{
 
 };
 
+struct AssociatedCalendar {
```
à quoi ça sert?

antoine-de replied:
ca embarque les différences par rapport au calendrier aussi

kinnou02 commented on source/type/type.proto:
```diff
@@ -291,6 +292,38 @@ message Address {
+}
+
+message CalendarException {
+    optional string datetime = 1;
```
pour le coup c'est plutot une date

antoine-de committed:
All: modification after long review :)

many improvement.

Note that it add some doc tests in python (but we need to be more
thorough in there coverage ^^)

antoine-de committed:
Protobuf: some review comments had been forgotten

antoine-de committed:
Jormungandr: disable doctest

I don't understand but with doc test the wrong .ini file is loaded thus
the tests do not pass on build bot.

We need to study that but for the moment I disable doctest

l-vincent-l commented on source/ed/connectors/fusio_parser.cpp:
```diff
@@ -689,6 +837,10 @@ void FusioParser::parse_files(Data& data) {
     parse<TripsFusioHandler>(data, "trips.txt", true);
     parse<StopTimeFusioHandler>(data, "stop_times.txt", true);
     parse<FrequenciesGtfsHandler>(data, "frequencies.txt");
+    parse<grid_calendar::GridCalendarFusioHandler>(data, "hcalendars.txt");
```
par curiosité pourquoi c'est préfixé par un  h ?

antoine-de replied:
aucune idée :)

ca doit etre changé mais il faut modifier fusio aussi du coup ca sera pour plus tard.

antoine-de committed:
Jormungandr: change the say the ini are loaded

to ease integration tests, we can define a INI_FILE in the settings to
load only one .ini file

l-vincent-l commented on source/ed/ed_persistor.cpp:
```diff
@@ -81,7 +81,7 @@ void EdPersistor::build_ways(){
     /// Add ways where admin is nil
     PQclear(this->lotus.exec("SELECT georef.add_fusion_ways();", "", PGRES_TUPLES_OK));
-    /// Update data in table 'rel_way_admin' by 'tmp_rel_way_admin'
+    /// Remplace data in table 'rel_way_admin' by 'tmp_rel_way_admin'
```
Sérieusement ? Une traduction en français d'un commentaire anglais ?

l-vincent-l commented on source/ed/ed_persistor.cpp:
```diff
@@ -821,6 +836,91 @@ void EdPersistor::insert_vehicle_journeys(const std::vector<types::VehicleJourne
+    this->lotus.prepare_bulk_insert("navitia.week_pattern", {"id", "monday", "tuesday", "wednesday", "thursday", "friday", "saturday", "sunday"});
+    std::vector<navitia::type::idx_t> to_insert;
+    for(const types::Calendar* cal : calendars){
+        navitia::type::idx_t idx = cal->week_pattern.to_ulong();
```
du coup c'est un id ou un uri non ?

l-vincent-l commented on source/ed/ed_persistor.cpp:
```diff
@@ -821,6 +836,91 @@ void EdPersistor::insert_vehicle_journeys(const std::vector<types::VehicleJourne
+    this->lotus.prepare_bulk_insert("navitia.calendar", {"id", "uri", "external_code", "name", "week_pattern_id"});
+
+    for(const types::Calendar* cal : calendars){
+        navitia::type::idx_t idx = cal->week_pattern.to_ulong();
```
même remarque

antoine-de committed:
Kraken: changes after review

l-vincent-l commented on source/jormungandr/jormungandr/interfaces/v1/ResourceUri.py:
```diff
@@ -94,13 +94,27 @@ def wrapper(*args, **kwargs):
                         "templated": templated
                     })
-            if collection in ['stop_areas', 'stop_points', 'addresses']:
+            if collection in ['stop_a', 'stop_points', 'addresses']:
```
petite boulette

l-vincent-l commented on source/jormungandr/jormungandr/scripts/default.py:
```diff
@@ -509,3 +510,7 @@ def pois(self, request, instance):
+
+    def calendars(self, request, instance):
+        return self.__on_ptref("calendars", type_pb2.CALENDAR,
+                               request, instance)
```
C'est quoi ce petit truc en fin ?
C'est parce qu'il n'y a pas de retour à la ligne ?

l-vincent-l commented on source/jormungandr/tests/check_utils.py:
```diff
@@ -0,0 +1,97 @@
+    return links
+
+
+def check_valid_calendar(cal):
```
C'est pas un utils pour tout le monde ça du coup non ?

l-vincent-l commented on source/jormungandr/tests/app_test.py:
```diff
@@ -24,31 +60,45 @@ def __init__(self, *args, **kwargs):
 
+        calendars = get_not_null(json_response, "calendars")
 
+        check_valid_calendar(calendars[0])
```
faudrait pas inverser les deux lignes ?

l-vincent-l commented on source/jormungandr/tests/check_utils.py:
```diff
@@ -0,0 +1,97 @@
+    tester = app.test_client(tester)
+    response = tester.get(url)
+
+    assert response is not None
```
Du coup c'est juste un assert non ?

l-vincent-l commented on source/jormungandr/tests/check_utils.py:
```diff
@@ -0,0 +1,97 @@
+        return val  # no check for booleans
+
+    assert val
+    return dict[field]
```
return val non ?

l-vincent-l commented on source/ptreferential/ptref_graph.cpp:
```diff
@@ -73,6 +73,10 @@
+
+    //from line to calendar
+    boost::add_edge(vertex_map[Type_e::Calendar], vertex_map[Type_e::Line], g);
+    boost::add_edge(vertex_map[Type_e::Line], vertex_map[Type_e::Calendar], g);
```
Du coup je n'ai pas suivi, vous avez gardé les calendar sur les lignes ?

antoine-de replied:
pour le moment oui. 
perso je manque de recul pour savoir ou faut le caler.

l-vincent-l commented on source/type/datetime.h:
```diff
@@ -83,4 +83,9 @@ inline std::string to_iso_string_no_fractional(Time t) {
 }
 
+/**
+ * weeday enum (because boost::date_time::weekdays start on sunday and we want monday to be the first day)
```
typo sur le weekday

l-vincent-l commented on source/type/response.proto:
```diff
@@ -341,6 +341,7 @@ message Response{
     repeated Route routes = 27;
     repeated Poi pois = 28;
     repeated PoiType poi_types = 29;
+    repeated Calendar calendars = 55;
```
30 non ?

l-vincent-l commented on source/type/type.cpp:
```diff
@@ -361,6 +382,7 @@ Mode_e static_data::modeByCaption(const std::string & mode_str) {
     case Type_e::Company: result.push_back(company->idx); break;
     case Type_e::PhysicalMode: result.push_back(journey_pattern->physical_mode->idx); break;
     case Type_e::ValidityPattern: result.push_back(validity_pattern->idx); break;
+        //TODO calendars
```
Du coup si c n'est pas fait ptref ne va pas fonctionner.

for(auto calendar : vehicle_journey->calendars) {
    result.push_back(calendar->idx);
}

antoine-de replied:
non car du coup on a bougé ca au niveau des lignes

antoine-de committed:
Build bot: try to make nosetest and doc test work for buildbot

antoine-de committed:
Jormungandr: changes after review

fix build bot

antoine-de committed:
for buildbot
